### PR TITLE
Inequality constraints in parameter space

### PR DESCRIPTION
**Pull request by marpulli**
Add ability to add inequality constraints to parameter space. This includes:
- Adding linear and non-linear inequality constraint objects as well as an IConstraint interface.
- Adding the scipy Trust Region Constrained optimizer that can optimize the acquisition in the presence of inequality constraints
- An `AnchorPointsGenerator` that will get anchor points that satisfy the constraints.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: emukit/core/constraints.py [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+        :param lower_bound: Lower bound vector of size (n_constraint,). Can be -np.inf for one sided constraint
+        :param upper_bound: Upper bound vector of size (n_constraint,). Can be np.inf for one sided constraint
+        :param jacobian_fun: Function describing
+        """
+        self.fun = fun
+        self.lower_bound = lower_bound
+        self.upper_bound = upper_bound
+        self.jacobian_fun = jacobian_fun if jacobian_fun else '2-point'
+
+    def evaluate(self, X: np.ndarray) -> np.ndarray:
```
Comment (mmahsereci) [inline emukit/core/constraints.py]: input should be small x to be more consistent.

File: emukit/core/constraints.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+import numpy as np
+
+_log = logging.getLogger(__name__)
+
+
+class IConstraint:
+    def evaluate(self, x: np.ndarray) -> np.ndarray:
+        """
+        :param x: Input locations to evaluate constraint at
```
Comment (mmahsereci) [inline emukit/core/constraints.py]: shape missing in all evaluate methods.

File: tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+@pytest.fixture
+def trust_region_constr_linear_constraint():
+    constraints = [LinearInequalityConstraint(np.array([[0, 1]]), np.array([0.5]), np.array([np.inf]))]
+    return OptTrustRegionConstrained([(-1, 1), (-1, 1)], constraints, 1000)
+
+
+def test_trust_region_constrained_no_context(trust_region_constr_linear_constraint, objective, space):
+    x0 = np.array([1, 1])
+    x, f = apply_optimizer(trust_region_constr_linear_constraint, x0, space, objective, None, None, None)
+    print(x)
```
Comment (mmahsereci) [inline tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py]: remove print statement

File: tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+    assert np.all(np.isclose(x, np.array([0, 0.5])))
+
+
+def test_trust_region_constrained_no_context(trust_region_constr_linear_constraint, objective, space):
+    x0 = np.array([1, 1])
+    x, f = apply_optimizer(trust_region_constr_linear_constraint, x0, space, objective, None, None, None)
+    print(x)
```
Comment (mmahsereci) [inline tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py]: remove print statement

File: emukit/core/constraints.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+                 constraint and zero if the constraint is violated
+        """
+        raise NotImplementedError
+
+
+class LinearInequalityConstraint(IConstraint):
+    """
+    Constraint of the form lower_bound < Ax < upper_bound where the matrix A is called "constraint_matrix"
+    """
+    def __init__(self, constraint_matrix: np.ndarray, lower_bound: np.ndarray, upper_bound: np.ndarray):
```
Comment (apaleyes) [inline emukit/core/constraints.py]: I think this could use some defaults. Say identity matrix, -infinity, +infinity. As well as validation, that A!=0 and at least one bound is set.

File: emukit/core/constraints.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    def evaluate(self, x: np.ndarray) -> np.ndarray:
+        """
+        :param x: Input locations to evaluate constraint at
+        :return: Numpy array of shape (n_input,) where an element will be 1 if the corresponding input satisfies the
+                 constraint and zero if the constraint is violated
+        """
+        raise NotImplementedError
+
+
+class LinearInequalityConstraint(IConstraint):
```
Comment (apaleyes) [inline emukit/core/constraints.py]: Isn't that just NonlinearInequalityConstraint with g(x) = Ax ? you could simplify the code a bit with this

File: emukit/core/constraints.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+        :param x: Input locations to evaluate constraint at
+        :return: Numpy array of shape (n_input,) where an element will be 1 if the corresponding input satisfies the
+                 constraint and zero if the constraint is violated
+        """
+        ax = self.constraint_matrix.dot(x)
+        return np.all([ax >= self.lower_bound, ax <= self.upper_bound], axis=0)
+
+
+class NonlinearInequalityConstraint:
```
Comment (apaleyes) [inline emukit/core/constraints.py]: if g(x) = Ax this becomes linear (see above). So how about just "InequalityConstraint" for the name?

File: emukit/core/constraints.py [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+    """
+    Constraint of the form lower_bound < g(x) < upper_bound
+    """
+    def __init__(self, fun: Callable, lower_bound: np.ndarray, upper_bound: np.ndarray,
```
Comment (apaleyes) [inline emukit/core/constraints.py]: again, i think defaults and validation for bounds would be good addition here

File: emukit/core/optimization/anchor_points_generator.py [hunk 9]
```diff
@@ -76,9 +76,36 @@ def __init__(self, space: ParameterSpace, acquisition: Acquisition, num_samples=
...
-    def get_anchor_point_scores(self, X) -> np.array:
+    def get_anchor_point_scores(self, X: np.ndarray) -> np.ndarray:
         """
         :param X: The samples at which to evaluate the criterion
         :return:
         """
         return self.acquisition.evaluate(X).flatten()
+
+
+class ConstrainedObjectiveAnchorPointsGenerator(AnchorPointsGenerator):
```
Comment (apaleyes) [inline emukit/core/optimization/anchor_points_generator.py]: Maybe we don't need a new class here? ObjectiveAnchorPointsGenerator is just ConstrainedObjectiveAnchorPointsGenerator with empty list of constraints, isn't it?

File: emukit/core/optimization/context_manager.py [hunk 10]
```diff
@@ -36,7 +36,7 @@ def __init__(self, space: ParameterSpace, context: Context):
             # Find encoded values of context variable
             param = self.space.get_parameter_by_name(name)
             if hasattr(param, 'encoding'):
-                self.context_values.append(param.encoding.get_encoding(context[name]))
+                self.context_values.append(int(param.encoding.get_encoding(context[name])))
```
Comment (apaleyes) [inline emukit/core/optimization/context_manager.py]: Why is this needed?

File: emukit/core/optimization/gradient_acquisition_optimizer.py [hunk 11]
```diff
@@ -67,3 +66,15 @@ def f_df(x):
 
         x_min, fx_min = min(optimized_points, key=lambda t: t[1])
         return x_min, -fx_min
+
+    def _get_anchor_points_generator(self, acquisition):
```
Comment (apaleyes) [inline emukit/core/optimization/gradient_acquisition_optimizer.py]: As noted above, we may be able to get rid of this getter

File: emukit/core/constraints.py (hunk 5 above)
Comment (marpulli) [reply emukit/core/constraints.py]: I'll add defaults for the bounds and some validation, I'm not keen on identity for constraint_matrix as you should be using the parameter bounds in that case rather than a constraint.

File: emukit/core/constraints.py (hunk 6 above)
Comment (marpulli) [reply emukit/core/constraints.py]: The optimizer will treat linear and non-linear constraints differently so it needs to know whether a certain constraint is linear or non-linear and they have to be represented differently. I don't think it's possible to merge the two inequality constraints.

Setting g(x) = Ax is possible but you would get lots of warnings from the optimizer and would be inefficient.

File: emukit/core/constraints.py (hunk 7 above)
Comment (marpulli) [reply emukit/core/constraints.py]: Same as above

File: emukit/core/constraints.py (hunk 8 above)
Comment (marpulli) [reply emukit/core/constraints.py]: I'll add them

File: emukit/core/optimization/anchor_points_generator.py (hunk 9 above)
Comment (marpulli) [reply emukit/core/optimization/anchor_points_generator.py]: Good point, I'll change this

File: emukit/core/optimization/context_manager.py (hunk 10 above)
Comment (marpulli) [reply emukit/core/optimization/context_manager.py]: My bad, it's not. I'll remove.

File: tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py (hunk 3 above)
Comment (marpulli) [reply tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py]: Fixed

File: emukit/core/constraints.py (hunk 2 above)
Comment (marpulli) [reply emukit/core/constraints.py]: Fixed

File: emukit/core/constraints.py (hunk 1 above)
Comment (marpulli) [reply emukit/core/constraints.py]: Fixed

File: tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py (hunk 4 above)
Comment (marpulli) [reply tests/emukit/core/optimization/test_trust_region_constrained_optimizer.py]: Fixed

File: emukit/core/constraints.py [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+        :param lower_bound: Lower bound vector of size (n_constraint,). Can be -np.inf for one sided constraint
+        :param upper_bound: Upper bound vector of size (n_constraint,). Can be np.inf for one sided constraint
+        """
+        if (lower_bound is None) and (upper_bound is None):
+            raise ValueError('Neither lower nor upper bounds is set, at least one must be specified')
+
+        # Default lower bound to -infinity
+        n_constraints = constraint_matrix.shape[0]
+        if lower_bound is None:
+            lower_bound = np.full([n_constraints], -np.nan)
```
Comment (apaleyes) [inline emukit/core/constraints.py]: Did you mean `np.inf` here and below?

File: emukit/core/constraints.py [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+    Constraint of the form lower_bound < Ax < upper_bound where the matrix A is called "constraint_matrix"
+    """
+    def __init__(self, constraint_matrix: np.ndarray, lower_bound: np.ndarray=None, upper_bound: np.ndarray=None):
+        """
+
+        :param constraint_matrix: (n_constraint, n_x_dims) matrix in b_lower < Ax < b_upper
+        :param lower_bound: Lower bound vector of size (n_constraint,). Can be -np.inf for one sided constraint
+        :param upper_bound: Upper bound vector of size (n_constraint,). Can be np.inf for one sided constraint
+        """
+        if (lower_bound is None) and (upper_bound is None):
```
Comment (apaleyes) [inline emukit/core/constraints.py]: You may also want to make sure they are not +-inf at the same time, maybe after defaults are handled

File: emukit/core/constraints.py (hunk 8 above)
Comment (apaleyes) [reply emukit/core/constraints.py]: Ok, I appreciate you need a way to differentiate inequalities. However current approach creates quite a bit of code repetition.

Maybe then we could have

```
class InequalityConstraint(IConstraint):

class LinearIC(InequalityConstraint):

class NonlinearIC(InequalityConstraint):
```

File: emukit/core/constraints.py [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+    Constraint of the form lower_bound < g(x) < upper_bound
+    """
+    def __init__(self, fun: Callable, lower_bound: np.ndarray, upper_bound: np.ndarray,
+                 jacobian_fun: Optional[Callable]=None):
+        """
+        :param fun: function defining constraint in b_lower < fun(x) < b_upper. Has signature f(x) -> array, shape(m,)
+                    where x is 1d and m is the number of constraints
+        :param lower_bound: Lower bound vector of size (n_constraint,). Can be -np.inf for one sided constraint
+        :param upper_bound: Upper bound vector of size (n_constraint,). Can be np.inf for one sided constraint
+        :param jacobian_fun: Function describing
```
Comment (apaleyes) [inline emukit/core/constraints.py]: describing...

Seems that is comes directly from scipy, can you add a reference here?

File: emukit/core/optimization/optimizer.py [hunk 15]
```diff
@@ -171,3 +173,70 @@ def f_df_no_context(self, x: np.array) -> Tuple[np.array, np.array]:
...
+
+        :param x0: Initial start point
+        :param f: Function to optimize
+        :param df: Derivative of function to optimize
+        :param f_df: Function returning both value of objective and its gradient
+        :return:
+        """
+        if df is None and f_df is not None:
+            df_1d = lambda x: f_df(x)[1]
+        elif df is not None:
```
Comment (apaleyes) [inline emukit/core/optimization/optimizer.py]: Some true magic happening here. What if `df` and `f_df` are both given, looks like this code will overwrite `f_df` then?

File: emukit/core/constraints.py (hunk 14 above)
Comment (apaleyes) [reply emukit/core/constraints.py]: Ok, agreement is not to tie this to scipy at all, so just expand the comment and move default string literal value for it to scipy conversion code inside the optimizer

File: emukit/core/constraints.py (hunk 13 above)
Comment (marpulli) [reply emukit/core/constraints.py]: Added that validation below

File: emukit/core/constraints.py (hunk 14 above)
Comment (marpulli) [reply emukit/core/constraints.py]: This is now done

File: emukit/core/constraints.py (hunk 12 above)
Comment (marpulli) [reply emukit/core/constraints.py]: Yep, this has been changed

File: emukit/core/optimization/gradient_acquisition_optimizer.py (hunk 11 above)
Comment (marpulli) [reply emukit/core/optimization/gradient_acquisition_optimizer.py]: Now removed

File: emukit/core/constraints.py (hunk 8 above)
Comment (marpulli) [reply emukit/core/constraints.py]: This has been done

**Comment by marpulli** [comment]
I've now addressed all comments and I think it's ready to merge - @apaleyes could you approve if you agree please?

File: tests/emukit/test_acquisitions.py [hunk 16]
```diff
@@ -98,6 +98,10 @@ def integrated_variance_acquisition(gpy_model, continuous_space):
 def squared_correlation_acquisition(vanilla_bq_model):
     return SquaredCorrelation(vanilla_bq_model)
 
+@pytest.fixture
+def probability_of_feasibility(gpy_model):
+    return ProbabilityOfFeasibility(gpy_model)
```
Comment (apaleyes) [inline tests/emukit/test_acquisitions.py]: I don't think we have this one merged yet, do we?

Comment (marpulli) [reply tests/emukit/test_acquisitions.py]: That's weird - I'm not sure how that got there. I've removed it
